zerver/tests/test_messages.py

```diff
@@ -2092,3 +2095,280 @@ def test_delete_message_second_time(self):
+        # already present in the UserMessage table. This test works on the
+        # fact that previous test just above this added a message but didn't
+        # updated the last_active_message_id field for the user.
+        sent_message = send_fake_message('Test Message 2', stream)
+        idle_user_msg_list = self.get_user_ums(long_term_idle_user)
+        idle_user_msg_count = len(idle_user_msg_list)
+        self.assertNotEqual(idle_user_msg_list[-1].message, sent_message)
+        with queries_captured() as queries:
+            add_missing_messages(long_term_idle_user)
+        self.assert_length(queries, 5)
+        idle_user_msg_list = self.get_user_ums(long_term_idle_user)
+        self.assertEqual(len(idle_user_msg_list), idle_user_msg_count + 1)
+        self.assertEqual(idle_user_msg_list[-1].message, sent_message)
+
+        # Test UserMessage rows are created correctly in case of stream
+        # Subscription was altered by admin while user was away.
+
+        # Test for a public stream.
+        sent_message_list = []
+        sent_message_list.append(send_fake_message('Test Message 3', stream))
+        # Alter subscription to stream by simply creating a log of the fact
+        # that the subscription was altered.
+        log_change_in_subscription(long_term_idle_user, stream, 'subscription_deactivated')
+        send_fake_message('Test Message 4', stream)
+        log_change_in_subscription(long_term_idle_user, stream, 'subscription_activated')
+        sent_message_list.append(send_fake_message('Test Message 5', stream))
+        sent_message_list.reverse()
+        idle_user_msg_list = self.get_user_ums(long_term_idle_user)
+        idle_user_msg_count = len(idle_user_msg_list)
+        for sent_message in sent_message_list:
+            self.assertNotEqual(idle_user_msg_list.pop().message, sent_message)
+        with queries_captured() as queries:
+            add_missing_messages(long_term_idle_user)
+        self.assert_length(queries, 5)
+        idle_user_msg_list = self.get_user_ums(long_term_idle_user)
+        self.assertEqual(len(idle_user_msg_list), idle_user_msg_count + 2)
+        for sent_message in sent_message_list:
+            self.assertEqual(idle_user_msg_list.pop().message, sent_message)
+
+        # Test for a Private Stream.
+        private_stream = self.make_stream('Core', invite_only=True)
+        self.subscribe_to_stream(self.example_email("iago"), "Core")
+        sent_message_list = []
+        send_fake_message('Test Message 6', private_stream)
+        self.subscribe_to_stream(self.example_email("hamlet"), "Core")
+        sent_message_list.append(send_fake_message('Test Message 7', private_stream))
+        log_change_in_subscription(long_term_idle_user, private_stream, 'subscription_deactivated')
+        send_fake_message('Test Message 8', private_stream)
+        log_change_in_subscription(long_term_idle_user, private_stream, 'subscription_activated')
```

Comment:
I think it might be better to actually change the user's subscriptions, rather than just logging changes; what do you think?

(Also, reading this made me think of a corner case: Do we correctly handle situations where you have a subscribe and an ubsubscribe with the same last message id?)